src/ast/ast_rewrite_same_clauses.c

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright Redis Ltd. 2018 - present
+ * Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ * the Server Side Public License v1 (SSPLv1).
+ */
+
+#include "ast.h"
+#include "../query_ctx.h"
+#include "../errors.h"
+#include "../util/arr.h"
+#include "../util/qsort.h"
+#include "../util/sds/sds.h"
+#include "../procedures/procedure.h"
+
+static void replace_create_clause
+(
+	cypher_astnode_t *root,      // ast root
+	cypher_astnode_t **clauses,  // clause being replaced
+	int scope_start,             // beginning of scope
+	int scope_end                // ending of scope
+) {
+	uint count = array_len(clauses);
+
+	struct cypher_input_range range = cypher_astnode_range(clauses[0]);
+	cypher_astnode_t **paths = array_new(cypher_astnode_t *, count);
+	for (uint i = 0; i < count; i++) {
+		const cypher_astnode_t *pattern = cypher_ast_create_get_pattern(clauses[i]);
+		uint npaths = cypher_ast_pattern_npaths(pattern);
+		for (uint j = 0; j < npaths; j++) {
+			const cypher_astnode_t *path = cypher_ast_pattern_get_path(pattern, j);
+			array_append(paths, cypher_ast_clone(path));	
+		}
+		cypher_ast_free(clauses[i]);
+	}
+	
+	//build the replacement pattern
+	cypher_astnode_t *pattern = cypher_ast_pattern(paths, array_len(paths), paths, array_len(paths), range);
+	
+	// build the replacement clause
+	cypher_astnode_t *new_clause = cypher_ast_create(false, pattern, &pattern, 1, range);
+
+	// replace original clause with the new one
+	cypher_ast_query_replace_clauses(root, new_clause, scope_start, scope_end);
```

Comment:
See if we can move the call to `cypher_ast_free` into `cypher_ast_query_replace_clauses`
freeing each compacted clause.
This way whoever calling `cypher_ast_query_replace_clauses` wouldn't need to remember freeing compacted clauses.